### PR TITLE
core(network-recorder): optimize network idle detection

### PR DESCRIPTION
**Pull request by brendankenny**
The `NetworkRecorder.recordsFromLogs` performance described [here](https://github.com/GoogleChrome/lighthouse/pull/9491#discussion_r321531671) was pretty bad, the vast majority of time spent in `findNetworkQuietPeriods`. Something like 1380ms spent on this is especially bad because it happens three times in a typical LH run: once [right before `GatherRunner.afterPass` ](https://github.com/GoogleChrome/lighthouse/blob/52cb0c904648be2f3395f2791b13844d7fac4d58/lighthouse-core/gather/gather-runner.js#L300), once in a computed artifact for the first audit that requests networkRecords for the defaultPass, and once spread over the run of `pass()` where `NetworkRecorder` is essentially running `recordsFromLogs` as protocol events come in. That 3x becomes 6x if e.g. there's a valid offline site served by a SW.

The biggest issue was constructing a full set of quiet periods for the entire set of network records when really for `networkIdle` and `network-2-idle` we only care if there's *currently* a quiet period. That's equivalent to [this check](https://github.com/GoogleChrome/lighthouse/blob/52cb0c904648be2f3395f2791b13844d7fac4d58/lighthouse-core/lib/network-recorder.js#L162-L165) passing so it adds a final period ending at `Infinity`.

This PR adds a new internal method `_isActiveIdlePeriod` that does essentially that (also catching the ignored network schemes check). Behavior is maintained exactly as before.

In testing I've found a 75-90% drop in execution time, the more requests on a page makes for a bigger performance win (so maybe taking 3 seconds off that LH run on @paulirish's machine). We could probably knock off another 25% in a few ways, but they should wait for follow ups as they're more invasive and need to be considered more carefully than this equivalent transformation.

### CODE REVIEW

**Comment by connorjclark** [comment]
I have a script (https://github.com/GoogleChrome/lighthouse/tree/timings-script) for measuring these things :)

image

Looks good!

**Comment by brendankenny** [comment]
> 

nice! Looks super handy.

**Comment by brendankenny** [comment]
Though if n is 30 for `lh:gather:getDevtoolsLog` that means it's treating the three calls per run as independent samples? That's going to throw things off since for most sites two of the three (offline and http redirect passes) are going to be basically zero. `lh:computed:NetworkRecords` should show the improvement more clearly, since we don't record computed artifact time when they're pulled from the cache.

The most straightforward fix to that might be to just sum them per run (so the number would be "the total amount of time spent in `lh:gather:getDevtoolsLog` for this run", not a record of individual calls) instead of treating them as independent samples.

**Comment by brendankenny** [comment]
cool :)

perf comparison

File: lighthouse-core/lib/network-recorder.js [hunk 1]
```diff
@@ -54,24 +54,37 @@ class NetworkRecorder extends EventEmitter {
...
-  _getActiveIdlePeriod(allowedRequests) {
-    const quietPeriods = NetworkRecorder.findNetworkQuietPeriods(this._records, allowedRequests);
-    return quietPeriods.find(period => !Number.isFinite(period.end));
+  _isActiveIdlePeriod(allowedRequests) {
+    let inflightRequests = 0;
+
+    for (let i = 0; i < this._records.length; i++) {
+      const record = this._records[i];
+      if (!NetworkRecorder.isNetworkRecordFinished(record)) {
+        if (IGNORED_NETWORK_SCHEMES.includes(record.parsedURL.scheme)) continue;
```
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/lib/network-recorder.js]: > nesting/mixing and matching seems a bit harder to parse IMO than consistency on one

sure, looks better

**Pull request merged**